### DISCUSSION TITLE
Consistent inline formatting with curly braces only

### DISCUSSION THREAD

**Discussion by rmnldwg** [jgm/djot#174] [Ideas]
First off, thanks for starting this project! It sounds super useful and promising to me.

While diving through this repo, the docs and the discussions, I had a thought: Since the syntax using openers and closers (e.g. `{_text_}` instead of `_text_`) resolves ambiguity and allows using many other punctuation characters for formatting (like discussed in #10), why not allow _only_ that syntax? Wouldn't it be super clean and consistent to do _all_ inline formatting using curly braces?

If I had to come up with a syntax, ignoring what I am used to from Markdown, it could be something like this: `{/emphasis/}`, `{*strong*}`, `{_underlined_}`, `{-deleted-}`, `{+inserted+}`, `{$ maths $}`, `` {`code`} ``, `{^superscript^}`, `{~subscript~}`, `{=highlighted=}`, `{,small caps,}`, ...

Would that be too obtrusive to read, compared to using the shorthand emphasis characters? If I look at the example below and squint a little, the `{}` are quite ignorable:

```djot
Here is a test paragraph with some {/emphasis/}. I am not sure,
but I think this is {*not too bad*}. Is it still readable
without too much {_distraction_}?
```

Another advantage is that at some point code editors could auto-close any `{char` opener with the appropriate `char}` closer.

Regarding maths, I think it may even be more intuitive and practical: With `` $`E=mc^2` `` the backticks indicate verbatim content, in contrast to `{$ E=mc^2 $}` where the syntax even allows to simply copy and paste equations from LaTeX and surround them with curly braces.

I'm curious to hear your thoughts on this. In any way, I'm a fan of djot already!

**Comment by jgm** [comment]
I've been tempted by this idea myself.

As for math, I think there'd still be an advantage to using verbatim backticks.  The problem with `{$..$}` is that tex math can contain the string `$}`, e.g. in `$x = \text{why not $y$}$` which is legal if odd.  So we start to get ambiguities, which are completely eliminated if we reuse verbatim syntax.

**Reply by rmnldwg** [reply]
Oh, I see! Thanks for clarifying :+1:

**Comment by waldyrious** [comment]
For prior art, Textile[^1] does this with `[]` (i.e. it supports `[**foobar**]` in addition to  `**foobar**`, and similarly for other markup), although it doesn't require it.

[^1]: I realize I may sound like a Textile fanboy as I've mentioned it in several of my comments here, but I did go through various lightweight markup languages looking for this feature, and except for those that use HTML-like tag names (e.g. [XMLite](https://github.com/lihaoyi/XMLite), [SexpCode](https://web.archive.org/web/20170523060320im_/cairnarvon.rotahall.org/misc/sexpcode.html), [PML](https://www.pml-lang.dev/), etc.), Textile was indeed the only one that seems to support this kind of delimited markup.

**Reply by gavinhoward** [reply]
Not the author, but thank you for the prior art.

My personal opinion is that braces are better than brackets because brackets are more likely to be used in actual text. (To adjust quotes, for example.)

**Reply by vassudanagunta** [reply]
It's important to keep in mind that most statements like "brackets are more likely to be used in actual text" tend to be English-centric.

**Reply by gavinhoward** [reply]
This is probably true. Can you tell me what languages use braces more?

**Reply by waldyrious** [reply]
To be clear, I wasn't advocating for brackets instead of braces :) I just wanted to give an example of a LML with broad support for such delimiters.

That said, I do think it would be nice if Djot consistently used brackets to wrap visible content (like the label of links, emphasised text, etc.), and braces to implicit content like classes and other attributes (or, dare I say, the URL of links :grin:). Or the other way around; I don't have a preference as I said above.

**Reply by vassudanagunta** [reply]
I don't know. I'm just saying if we want a markup to not be English-centric, you can't make statements like that without analysis. There is wide variation in how different languages do things, e.g. quotations: `« Le singe est sur la branche ! »` vs `„Der Affe ist auf dem Ast!‟` vs `「樹枝上的猴子！」`.  Chinese also uses `《》〈〉` for titles and quotations, though they are full width so distinguishable from `<>` (I don't know if Chinese writers use the latter for convenience). Also `「...」` is also used in Chinese for emphasis so no additional markup is needed.

It's very easy to neglect this since English dominates even the discussion of the design of plain text formats.

I wonder if formats like Markdown are de facto English or at least European-language oriented and if there simply should be other markup languages for other language families.

**Reply by bpj** [reply]
@vassudanagunta I’m not sure if I’m with you here. In Chinese you simply don’t need emphasis in the Western sense because punctuation already takes care of it but in mixed-language texts it comes back in if you have quotations in Western scripts. I'm all with you WRT international quote marks: smart quotes should be configurable!

As for the original proposal I tend to be in favor of it since it removes most ambiguity. Not least intra-word underscores and such things as my old discipline’s use of literal asterisks before/after words, which i *never* remember to escape the first time around, would become non-problematic and uniformity would increase.

I agree that curlies probably are better for balanced delimiters than square brackets but I think that ideally brackets for delimiters, spans, links, attributes etc. and *at least* smart quotes — in Pandoc we who use Western languages other than English use filters to localize quotes! — should be configurable. Smart quotes are probably best configured in metadata on a per-document basis while delimiters for the parser probably should be configured in a table of directives passed to the parser by the host program. The only requirement should be that there should be a delimiter pair for each kind, and that each delimiter is unique. People will be punished by themselves if they choose too weird delimiters.

One thing that worries me about djot is that both balanced delimiters and attributes use curlies, since that prevents using `. # %` for balanced delimiters! :-) Part of me wishes that any (at least any non-paired) punctuation characters could be used as balanced delimiters and stored in the AST in such a way that renderers could do whatever they please with them.

In my homegrown LML which I used to generate LaTeX before I discovered Pandoc I actually used `{*textbf*} {_textit_} {/emph/}` etc. with always required curlies, and later I wrote a “preprocessor” which replaced/wrapped chunks of text inside possibly nested delimiters (using perl \>= 5.10.1 regexes[^1]) with/in arbitrary markup/texts, taking configuration from a YAML file. In short it faked Pandoc’s span and div syntax and then some before Pandoc supported it, including the possibility to interpolate argument values and other predefined data in the replacement text with [String::Formatter](https://metacpan.org/pod/String::Formatter). I still sometimes find this tool easier to use than a ton of bespoke Pandoc filters.

[^1]: Not the world’s most efficient regexes in that they use things like `(?s:(?!$open|$close).)+` but I never use this tool in anything time critical!

**Reply by vassudanagunta** [reply]
> I’m not sure if I’m with you here.

I think you're 100% with me. My point wasn't that `{}` shouldn't be used, but 

>  if we want a markup to not be English-centric, you can't make statements like ["brackets are more likely to be used in actual text"] without analysis.

You followed up with some analysis, even if you didn't show that using `{}` for markup won't collide with its use as natural punctuation in other languages. I have a feeling that braces are safe to use, but that's a feeling, anecdotal, not data. The rest of my comment was only providing examples of how an English-centric view can result in markup design that won't work for many other languages. 

As to your thoughts about configurability, I'd measure their alignment with djot's goals.

**Comment by itraveller1** [comment]
The idea is very tempting, but unfortunately ergonomics and unambiguity are at different poles of syntax. Rejection of natural markers will move Djot towards a declarative syntax, where everything is unambiguous, but not very convenient.

A healthy balance is probably needed when the necessary minimum of markers is maintained, which expresses the most requested part of the handwriting syntax. These markers, of course, must be duplicated in a declarative style, like the rest of the functionality. I think @jgm did a very good job of that.

**Comment by dereuromark** [comment]
### How Non-Technical Users Mark Up Text

Observing how people annotate paper documents reveals natural instincts:

| Intent          | Paper Action           | Keyboard Approximation |
|-----------------|------------------------|------------------------|
| Emphasis        | Underline              | `_text_`               |
| Strong emphasis | Circle / Box           | `[text]` or `*text*`   |
| Deletion        | Strikethrough          | `~text~` or `-text-`   |
| Insertion       | Caret + write above    | `^text^`               |
| Comment         | Margin note            | `// comment`           |
| Reference       | Number in circle       | `[1]` or `(1)`         |
| Quote           | Quote marks            | `"text"` or `> text`   |

### What People Get Wrong in Markdown

From teaching Markdown to non-programmers:

1. **Link syntax order** - "Is it `[]()` or `[]()`?" (yes, they ask this)
2. **Nested lists** - Not understanding the indent requirements
3. **Code blocks** - Backtick key location varies by keyboard
4. **Line breaks** - Two spaces at end of line is invisible
5. **Emphasis** - "Do I use one or two asterisks?"
6. **Escaping** - Not knowing why their `*` became italic

### The "Five-Second Rule"

A good syntax should be:
- **Learnable in 5 seconds** for basic use
- **Memorable after 5 days** without use
- **Unambiguous within 5 characters** of context


So I agree that a good balance of non-tech friendly and nerdy is probably in order :)
The current default for strong/emphasis only seems to balance this.

Personally, I would also like to have a few more without the harder to type extra {}, but fair enough.

**Reply by Omikhleia** [reply]
N.B.
> Strong emphasis ... Circle / Box
> Deletion ... Strikethrough

Formally insertions/deletions are not semantically linked to strike-through. I use all three for some documents with critical apparatus.
I don't know where in the world strong emphasis is related to a "circle or box", or where "comments" could become margin notes.

And you forgot highlighting (useful in some contexts too), super-scripting and sub-scripting. 
So this is a very "ad hoc" list, but the so-said "natural instincts" do not seem to match any real-world usages and needs (and I don't even intend to be negative here, I just don't understand where these definitions lead to -- but Djot offers all of the above currently, out-of-the-box, without those weird re-definitions). :cat:

**Comment by dereuromark** [comment]
@rmnldwg I went with your initial proposal in [Carve](https://markup-carve.github.io/carve/).
I think you got a point here regarding intuitive behavior.
So I stand corrected on my initial review last year.

But small caps can be done using a class and css. So it should not have its own form IMO.
